<a href="https://colab.research.google.com/github/codedyasai/Python_MachineLearning/blob/main/14_XGB_%ED%99%95%EC%9D%B8%ED%95%99%EC%8A%B5_%EB%8D%B0%EC%9D%B4%ED%8C%85%EC%95%B1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from IPython.display import display
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
!pip install mglearn
import mglearn

# 음수표현 라이브러리
plt.rcParams['axes.unicode_minus'] = False

# 경고무시
import warnings
warnings.filterwarnings("ignore")

# 매직명령어 : 시각화 결과가 노트북에 포함되도록
%matplotlib inline

# 데이터셋과 주제 정의
- match의 성사 여부 추측 (분류)
- 평가지표: 정확도, 혼동행렬, 분류 리포트

In [ ]:
dating = pd.read_csv('speed_dating.csv', index_col= 0)
dating.head()

has_null  gender   age  age_o                                  race  \
0         0  female  21.0   27.0  Asian/PacificIslander/Asian-American   
1         0  female  21.0   22.0  Asian/PacificIslander/Asian-American   
2         1  female  21.0   22.0  Asian/PacificIslander/Asian-American   
3         0  female  21.0   23.0  Asian/PacificIslander/Asian-American   
4         0  female  21.0   24.0  Asian/PacificIslander/Asian-American   

                                 race_o  importance_same_race  \
0           European/Caucasian-American                   2.0   
1           European/Caucasian-American                   2.0   
2  Asian/PacificIslander/Asian-American                   2.0   
3           European/Caucasian-American                   2.0   
4               Latino/HispanicAmerican                   2.0   

   importance_same_religion  pref_o_attractive  pref_o_sincere  ...  \
0                       4.0               35.0            20.0  ...   
1                       4.0               60.0             0.0  ...   
2                       4.0               19.0            18.0  ...   
3                       4.0               30.0             5.0  ...   
4                       4.0               30.0            10.0  ...   

   funny_partner  ambition_partner  shared_interests_partner  \
0            7.0               6.0                       5.0   
1            8.0               5.0                       6.0   
2            8.0               5.0                       7.0   
3            7.0               6.0                       8.0   
4            7.0               6.0                       6.0   

   interests_correlate  expected_happy_with_sd_people  \
0                 0.14                            3.0   
1                 0.54                            3.0   
2                 0.16                            3.0   
3                 0.61                            3.0   
4                 0.21                            3.0   

   expected_num_interested_in_me  like  guess_prob_liked  met  match  
0                            2.0   7.0               6.0  0.0      0  
1                            2.0   7.0               5.0  1.0      0  
2                            2.0   7.0               NaN  1.0      1  
3                            2.0   7.0               6.0  0.0      1  
4                            2.0   6.0               6.0  0.0      1  

[5 rows x 39 columns]

In [ ]:
dating.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8378 entries, 0 to 8377
Data columns (total 39 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   has_null                       8378 non-null   int64  
 1   gender                         8378 non-null   object 
 2   age                            8283 non-null   float64
 3   age_o                          8274 non-null   float64
 4   race                           8315 non-null   object 
 5   race_o                         8305 non-null   object 
 6   importance_same_race           8299 non-null   float64
 7   importance_same_religion       8299 non-null   float64
 8   pref_o_attractive              8289 non-null   float64
 9   pref_o_sincere                 8289 non-null   float64
 10  pref_o_intelligence            8289 non-null   float64
 11  pref_o_funny                   8280 non-null   float64
 12  pref_o_ambitious               8271 non-null   f

In [ ]:
dating.describe()

has_null          age        age_o  importance_same_race  \
count  8378.00000  8283.000000  8274.000000           8299.000000   
mean      0.87491    26.358928    26.364999              3.784793   
std       0.33084     3.566763     3.563648              2.845708   
min       0.00000    18.000000    18.000000              0.000000   
25%       1.00000    24.000000    24.000000              1.000000   
50%       1.00000    26.000000    26.000000              3.000000   
75%       1.00000    28.000000    28.000000              6.000000   
max       1.00000    55.000000    55.000000             10.000000   

       importance_same_religion  pref_o_attractive  pref_o_sincere  \
count               8299.000000        8289.000000     8289.000000   
mean                   3.651645          22.495347       17.396867   
std                    2.805237          12.569802        7.044003   
min                    1.000000           0.000000        0.000000   
25%                    1.000000          15.000000       15.000000   
50%                    3.000000          20.000000       18.370000   
75%                    6.000000          25.000000       20.000000   
max                   10.000000         100.000000       60.000000   

       pref_o_intelligence  pref_o_funny  pref_o_ambitious  ...  \
count          8289.000000   8280.000000       8271.000000  ...   
mean             20.270759     17.459714         10.685375  ...   
std               6.782895      6.085526          6.126544  ...   
min               0.000000      0.000000          0.000000  ...   
25%              17.390000     15.000000          5.000000  ...   
50%              20.000000     18.000000         10.000000  ...   
75%              23.810000     20.000000         15.000000  ...   
max              50.000000     50.000000         53.000000  ...   

       funny_partner  ambition_partner  shared_interests_partner  \
count    8028.000000       7666.000000               7311.000000   
mean        6.400598          6.777524                  5.474559   
std         1.953702          1.794055                  2.156363   
min         0.000000          0.000000                  0.000000   
25%         5.000000          6.000000                  4.000000   
50%         7.000000          7.000000                  6.000000   
75%         8.000000          8.000000                  7.000000   
max        10.000000         10.000000                 10.000000   

       interests_correlate  expected_happy_with_sd_people  \
count          8220.000000                    8277.000000   
mean              0.196010                       5.534131   
std               0.303539                       1.734059   
min              -0.830000                       1.000000   
25%              -0.020000                       5.000000   
50%               0.210000                       6.000000   
75%               0.430000                       7.000000   
max               0.910000                      10.000000   

       expected_num_interested_in_me         like  guess_prob_liked  \
count                    1800.000000  8138.000000       8069.000000   
mean                        5.570556     6.134087          5.207523   
std                         4.762569     1.841285          2.129565   
min                         0.000000     0.000000          0.000000   
25%                         2.000000     5.000000          4.000000   
50%                         4.000000     6.000000          5.000000   
75%                         8.000000     7.000000          7.000000   
max                        20.000000    10.000000         10.000000   

               met        match  
count  8003.000000  8378.000000  
mean      0.049856     0.164717  
std       0.282168     0.370947  
min       0.000000     0.000000  
25%       0.000000     0.000000  
50%       0.000000     0.000000  
75%       0.000000     0.000000  
max       8.000000     1.000000  

[8 rows x 36 columns]

In [ ]:
dating.has_null.value_counts()

1    7330
0    1048
Name: has_null, dtype: int64

In [ ]:
dating.isnull().mean()

has_null                         0.000000
gender                           0.000000
age                              0.011339
age_o                            0.012413
race                             0.007520
race_o                           0.008713
importance_same_race             0.009429
importance_same_religion         0.009429
pref_o_attractive                0.010623
pref_o_sincere                   0.010623
pref_o_intelligence              0.010623
pref_o_funny                     0.011697
pref_o_ambitious                 0.012772
pref_o_shared_interests          0.015397
attractive_o                     0.025304
sincere_o                        0.034256
intelligence_o                   0.036524
funny_o                          0.042970
ambitous_o                       0.086178
shared_interests_o               0.128432
attractive_important             0.009429
sincere_important                0.009429
intellicence_important           0.009429
funny_important                  0

- 결측치가 지나치게 많은 expected_num_interested_in_me,
- 의미가 없어질 has_null을 지우자.

In [ ]:
dating.drop(['has_null', 'expected_num_interested_in_me'], axis= 1, inplace= True)
dating.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8378 entries, 0 to 8377
Data columns (total 37 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   gender                         8378 non-null   object 
 1   age                            8283 non-null   float64
 2   age_o                          8274 non-null   float64
 3   race                           8315 non-null   object 
 4   race_o                         8305 non-null   object 
 5   importance_same_race           8299 non-null   float64
 6   importance_same_religion       8299 non-null   float64
 7   pref_o_attractive              8289 non-null   float64
 8   pref_o_sincere                 8289 non-null   float64
 9   pref_o_intelligence            8289 non-null   float64
 10  pref_o_funny                   8280 non-null   float64
 11  pref_o_ambitious               8271 non-null   float64
 12  pref_o_shared_interests        8249 non-null   f

In [ ]:
# 결측치가 없는 행들의 수
dating.notna().all(axis=1).sum()

5635

- 5600개 가량이다. 너무 많이 버리는 것일지도 모르지만, 이것들만을 사용해보자.

In [ ]:
data = dating.drop(columns='match')
target = dating.match

In [ ]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(data, target,
                                                    test_size= 0.2, random_state= 42)

---------------------------------------------------------------------------

In [ ]:
df = pd.read_csv('speed_dating.csv', index_col= 0)
df.head()

has_null  gender   age  age_o                                  race  \
0         0  female  21.0   27.0  Asian/PacificIslander/Asian-American   
1         0  female  21.0   22.0  Asian/PacificIslander/Asian-American   
2         1  female  21.0   22.0  Asian/PacificIslander/Asian-American   
3         0  female  21.0   23.0  Asian/PacificIslander/Asian-American   
4         0  female  21.0   24.0  Asian/PacificIslander/Asian-American   

                                 race_o  importance_same_race  \
0           European/Caucasian-American                   2.0   
1           European/Caucasian-American                   2.0   
2  Asian/PacificIslander/Asian-American                   2.0   
3           European/Caucasian-American                   2.0   
4               Latino/HispanicAmerican                   2.0   

   importance_same_religion  pref_o_attractive  pref_o_sincere  ...  \
0                       4.0               35.0            20.0  ...   
1                       4.0               60.0             0.0  ...   
2                       4.0               19.0            18.0  ...   
3                       4.0               30.0             5.0  ...   
4                       4.0               30.0            10.0  ...   

   funny_partner  ambition_partner  shared_interests_partner  \
0            7.0               6.0                       5.0   
1            8.0               5.0                       6.0   
2            8.0               5.0                       7.0   
3            7.0               6.0                       8.0   
4            7.0               6.0                       6.0   

   interests_correlate  expected_happy_with_sd_people  \
0                 0.14                            3.0   
1                 0.54                            3.0   
2                 0.16                            3.0   
3                 0.61                            3.0   
4                 0.21                            3.0   

   expected_num_interested_in_me  like  guess_prob_liked  met  match  
0                            2.0   7.0               6.0  0.0      0  
1                            2.0   7.0               5.0  1.0      0  
2                            2.0   7.0               NaN  1.0      1  
3                            2.0   7.0               6.0  0.0      1  
4                            2.0   6.0               6.0  0.0      1  

[5 rows x 39 columns]

- 최대값이 특이한 부분을 확인
-본인 및 상대방을 평가하는 변수(xxx_o, xxx_partner)는 최대값이 10으로 일정해보인다.
- 중요도와 관련된 변수(pred_o_xxx, xxx_important)는 최대값이 100부터 60, 50, ...
- 평가 관련된 변수는 대체적으로 0-10으로 개별 항목에 적용
- 중요도 관련변수는 최대값 100점을 각 항목에 분배해서 적용




**결측치 처리**
- 일반적으로 모든 변수의 결측치가 5% 미만
- 숫자데이터는 범위를 벗어나거나 일반적으로 나올법하지 않은 임의의 숫자는 -99로 대체
- 중요도와 관련된 변수들은 결측치를 제거
- 중요도 * 점수
- dropna() 결측치 제거

## 일부 feature의 결측치 제거
- dropna()
- 중요도와 관련된 변수들의 결측치 제거

In [ ]:
df = df.dropna(subset=['pref_o_attractive',
       'pref_o_sincere', 'pref_o_intelligence', 'pref_o_funny',
       'pref_o_ambitious', 'pref_o_shared_interests', 'attractive_important', 'sincere_important',
       'intellicence_important', 'funny_important', 'ambtition_important',
       'shared_interests_important'])

- 나머지 변수에 존재하는 결측치는 -99로 대체

In [ ]:
df = df.fillna(-99)

## 피처 엔지니어링

- 파생변수를 생성하거나 변수들을 이용하여 계산을 수행하는 것처럼 각 피처들의 값을 조정

In [ ]:
# age를 처리할 목적의 함수를 구현
def age_gap(x):
    if x['age'] == -99:
        return -99
    elif x['age_o'] == -99:
        return -99
        # 연상 또는 연하
    elif x['gender'] == 'female':
        return x['age_o'] - x['age']
    else:
        return x['age']  - x ['age_o']

- 남녀중 한명이라도 -99면 -99를 반환
- 남자가 연상이면 플러스, 여자가 연상이면 마이너스값 반환

In [ ]:
df['age_gap'] = df.apply(age_gap, axis= 1)
# age_gap 변수에 절대값을 이용하여 나이차에 순수 나이차를 담는다.
df['age_gap_abs'] = abs(df['age_gap'])

- 본인과 상대방이 인종이 같으면 1, 다르면 -1로 처리, 결측치는 -99로 처리

In [ ]:
def same_race(x):
    if x['race'] == -99:
        return -99
    elif x['race_o'] == -99:
        return -99
    elif x['race'] == x['race_o']:
        return 1
    else:
        return -1

In [ ]:
df['same_race'] = df.apply(same_race, axis= 1)

In [ ]:
def same_race_point(x):
    if x['same_race'] == -99:
        return -99
    else:
        return x['same_race'] * x['importance_same_race']

In [ ]:
df['same_race_point'] = df.apply(same_race_point, axis= 1)

- 평가항목과 중요도를 곱해서 파생변수 생성

In [ ]:
def rating(data, importance, score):
    if data[importance] == -99:
        return -99
    elif data[score] == -99:
        return -99
    else:
        return data[importance] * data[score]

In [ ]:
# 각 항목에 대한 변수들을 리스트로 생성
## 상대방의 중요도
partner_imp = df.columns[8:14]
## 나에 대한 상대방의 평가
partner_rate_me = df.columns[14:20]
## 나의 중요도
my_imp = df.columns[20:26]
## 상대방에 대한 나의 평가
my_rate_partner = df.columns[26:32]

- 각 항목(6가지)에 대한 파생변수 생성
- 항목별 점수(중요도 * 평가점수)를 담는다.
- 'attractive_p','sincere_p', 'intelligence_p', 'funny_p', 'ambitous_p', 'shared_interests_p'

In [ ]:
# 상대방 관련 점수를 담을 변수 리스트
new_label_partner = ['attractive_p','sincere_p', 'intelligence_p', 'funny_p', 'ambitous_p', 'shared_interests_p']

# 나의 관련 점수를 담을 리스트
new_label_me = ['attractive_m','sincere_m', 'intelligence_m', 'funny_m', 'ambitous_m', 'shared_interests_m']

In [ ]:
for i,j,k in zip(new_label_partner, partner_imp, partner_rate_me):
    df[i] = df.apply(lambda x: rating(x,j,k), axis= 1)

In [ ]:
for i,j,k in zip(new_label_me, my_imp, my_rate_partner):
    df[i] = df.apply(lambda x: rating(x,j,k), axis= 1)

- object 타입에 변수들 gender, race, race_o

In [ ]:
df = pd.get_dummies(df, columns=['gender', 'race', 'race_o'], drop_first= True)
# drop_first: 첫번째 컬럼 삭제

In [ ]:
df.head()

has_null   age  age_o  importance_same_race  importance_same_religion  \
0         0  21.0   27.0                   2.0                       4.0   
1         0  21.0   22.0                   2.0                       4.0   
2         1  21.0   22.0                   2.0                       4.0   
3         0  21.0   23.0                   2.0                       4.0   
4         0  21.0   24.0                   2.0                       4.0   

   pref_o_attractive  pref_o_sincere  pref_o_intelligence  pref_o_funny  \
0               35.0            20.0                 20.0          20.0   
1               60.0             0.0                  0.0          40.0   
2               19.0            18.0                 19.0          18.0   
3               30.0             5.0                 15.0          40.0   
4               30.0            10.0                 20.0          10.0   

   pref_o_ambitious  ...  shared_interests_m  gender_male  \
0               0.0  ...                75.0            0   
1               0.0  ...                90.0            0   
2              14.0  ...               105.0            0   
3               5.0  ...               120.0            0   
4              10.0  ...                90.0            0   

   race_Black/AfricanAmerican  race_European/Caucasian-American  \
0                           0                                 0   
1                           0                                 0   
2                           0                                 0   
3                           0                                 0   
4                           0                                 0   

   race_Latino/HispanicAmerican  race_Other  race_o_Black/AfricanAmerican  \
0                             0           0                             0   
1                             0           0                             0   
2                             0           0                             0   
3                             0           0                             0   
4                             0           0                             0   

   race_o_European/Caucasian-American  race_o_Latino/HispanicAmerican  \
0                                   1                               0   
1                                   1                               0   
2                                   0                               0   
3                                   1                               0   
4                                   0                               1   

   race_o_Other  
0             0  
1             0  
2             0  
3             0  
4             0  

[5 rows x 61 columns]

# 모델링과 평가

## 훈련세트와 테스트세트 분리

In [ ]:
from sklearn.model_selection import train_test_split

# x: feature
# y: target

x_train, x_test, y_train, y_test = train_test_split(df.drop('match', axis= 1),
                                                    df['match'], test_size= 0.2, random_state= 100)

## XGBoost 모델 구현

In [ ]:
import xgboost as xgb

In [ ]:
model = xgb.XGBClassifier(n_estimators= 500, max_depth= 5, random_state= 100)
model.fit(x_train, y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=5, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=500, n_jobs=None,
              num_parallel_tree=None, random_state=100, ...)

In [ ]:
pred = model.predict(x_test)

print(model.score(x_train, y_train))
print(model.score(x_test, y_test))

1.0
0.8690036900369004


In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
accuracy_score(y_test, pred)

0.8690036900369004

In [ ]:
print(confusion_matrix(y_test, pred))

[[1300   65]
 [ 148  113]]


- 1종오류: 65개, 실제는 0인데 1로 예측
- 2종오류: 148개, 실제는 1인데 0으로 예측

In [ ]:
print(classification_report(y_test, pred))
# 1에 대한 것만 봐야함

              precision    recall  f1-score   support

           0       0.90      0.95      0.92      1365
           1       0.63      0.43      0.51       261

    accuracy                           0.87      1626
   macro avg       0.77      0.69      0.72      1626
weighted avg       0.86      0.87      0.86      1626



0.63, 0.43, 0.51